src/emmy/expression/cse.cljc

```diff
@@ -7,152 +7,154 @@
+;; subexpressions of a commutative operation.
+;;
+;; We begin by rewriting our input expression to avoid this. While this
+;; will likely generate a tree with more nodes, it exposes each primitive
+;; computation to the algorithm which enables a maximal factoring (at the
+;; cost, as we will see, of generating code in which operations are limited
+;; to two operands. This may produce less compact-looking compiled code,
+;; but should be perfectly transparent to the code generators of the host
+;; languages.)
+
+(defn- dissociate
+  "If `x` is an expression like `(f x1 x2 x3...)`, where `(f? f)` is true,
+  returns `(f ... (f x1 x2) x3 ...)`, i.e., we replace applications with
+  arity > 2 with left-associative nested applications of arity 2. Otherwise
+  `x` is returned."
+  [f? x]
+  (letfn [(go [x]
+              (cond
+                (not (seq? x)) x
+                (not (f? (first x))) (map go x)
+                (< (count x) 4) (map go x)
+                :else (let [[f & xs] x]
+                        (reduce (partial list f) (map go xs)))))]
+    (go x)))
+
+;; This is the heart of the F90 algorithm. We do a post-order traversal
+;; of the tree, building an association between equivalence classes of
+;; subtree to small integers. We can then represent the maximally factored
+;; DAG as a simple sequence of expressions, where integers in argument
+;; position represent back-references to the indexed computation emitted
+;; previously (that is, in the nth expression, indices in the range [0..n-1]
+;; may appear). Thus the computation may be efficiently performed by
+;; evaluation the subexpressions in the order reported.
+
+(defn- uid-assigner
+  "Produces a function which takes an S-expression and produces the
+  compacted DAG representation of the tree, as described in [Flajolet 90].
+  Returns the ID of the root of the tree. The assigner generated by this
+  function can be used multiple times, and earlier compact forms will be reused
+  by new expressions; in this way, multiple expressions in the same scope
+  may be compacted.
+
+  Passing the argument `:table` will result in a sequence
+  of simple forms with a function symbol at the head followed by arguments.
+  The arguments are either symbols drawn from the original expression or
+  integers which represent the the index of the value of a computation
+  performed earlier in the sequence."
+  []
+  (let [counter (atom -1)
+        table   (atom {})]
+    (letfn [(install [x]
+              (or (@table x)
+                  (let [i (swap! counter inc)]
+                    (swap! table assoc x i)
+                    i)))
+            (uid [x]
+                (cond
+                  (symbol? x) x
+                  (number? x) (install x)
+                  (seq? x) (install (into [(first x)] (map uid (next x))))
+                  :else (throw (u/exception (str "cse? " x)))))]
+      (fn [expr]
+        (if (= expr :table)
+          (->> @table (sort-by second) (map first))
```

Comment:
this is another case where I will recommend a `doall`, given that this is querying a stateful data structure and we probably want the query to trigger at the time of the call.

Reply:
But the value of `@table` is immutable?

Reply:
oh, yeah, good call! Told you I had paranoia.